Improve goToSiteEditor E2E utils

tomasztunik opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This is a follow up to #6080 as discussed [here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6080#discussion_r833195402). 

The change cleans up `tests/e2e/specs/backend/site-editing-templates.test.js` E2E tests by hiding UX details of waiting for navigation to different contexts of Site Editor inside the goToSiteEditor E2E. 

It will: 
- wait for table being rendered when accessing template or template parts list - via `postType` param (defaults to `wp_template`)
- or waiting for Actions controls on template or template parts list when we want to ensure latest templates state is fetched and Actions are available on modified templates - via `waitForActions` param.
- wait for canvas when going to editing a template or template_part view - via `postId` param

By default the method if no params are provided will take us to the templates list view.

<!-- Reference any related issues or PRs here -->
Fixes #6080 

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests
### Manual Testing

How to test the changes in this Pull Request:

1. E2E CI should pass

tomasztunik committed:
dry getAllTemplates util

tomasztunik committed:
Improve goToSiteEditor e2e util

It will now properly wait for page to load
and critical parts to be available before
handing control over to the next steps

tomasztunik committed:
adapt site editing tests to new goToSiteEditor API

tomasztunik committed:
try to stabilize the saveTemplate

to early navigation was causing the dreaded
"Expected an Error, but "" was thrown"

tomasztunik committed:
try stabilizing saveTemplate

tomasztunik committed:
contextually group/sort params

sunyatasattva reviewed (commented):
I'm personally not a big fan of a function that does too many things depending on which parameters are passed to it.

Puppeteer itself had a function called `waitFor` which would behave differently depending on what arguments were passed. They have since deprecated that in favor of more specific functions.

I agree with this approach, so instead of a `goToSiteEditor` that behaves differently depending on where you want to go, I'd have a `goToTemplatesList` that does just that: you can still call `goToSiteEditor` under the hood and then basically just add the `waitForSelector`.

I'm not putting this as a “requesting change“, because I'd love to hear your opinion. If you are strongly married to this approach, I will approve it no problem.

tomasztunik commented:
I'd see it all as a PageObject and thought it could be improvement for the
time being. This method was already handling the same cases it just wasn't
taking full responsibility for hiding the UX/waiting details you mentioned.

If you think this will be an improvement for now I'd go with it and then do
a proper move to pageobjects once we are there and split the responsibility
better. At least it would be all now in one place making it much easier I
think.

A to-do comment inside?

pt., 25 mar 2022, 17:31 użytkownik Lucio Giannotta ***@***.***>
napisał:

> ***@***.**** commented on this pull request.
>
> I'm personally not a big fan of a function that does too many things
> depending on which parameters are passed to it.
>
> Puppeteer itself had a function called waitFor which would behave
> differently depending on what arguments were passed. They have since
> deprecated that in favor of more specific functions.
>
> I agree with this approach, so instead of a goToSiteEditor that behaves
> differently depending on where you want to go, I'd have a
> goToTemplatesList that does just that: you can still call goToSiteEditor
> under the hood and then basically just add the waitForSelector.
>
> I'm not putting this as a “requesting change“, because I'd love to hear
> your opinion. If you are strongly married to this approach, I will approve
> it no problem.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6109#pullrequestreview-921884828>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA3NDQ47HGQCBFWWKRDJFDVBXS7JANCNFSM5RO6UJLQ>
> .
> You are receiving this because you authored the thread.Message ID:
> <woocommerce/woocommerce-gutenberg-products-block/pull/6109/review/921884828
> @github.com>
>

sunyatasattva reviewed (approved):
I'm approving, for now, then. I don't think it's worth to add a todo comment: this is already a non blocking issue which sprang from a todo comment. The fix to this is fast enough; you know what they say, _if it takes less than 2 minutes, don't create a todo, do it now!_

So, this for me is an architecture decision: if we go with this current implementation, it is because we like it. If we think separation of concerns of this util is useful/better, we should just do it now because it takes very little time.

I'm ok with both, but I wouldn't go on spawning tickets.

tomasztunik commented:
Let's split it then - it's the solution we've originally discussed.

Should we split it by type or by context? Ie. `goToTemplatePart` vs `goToTemplate` with ability to go to list or item view by providing id and hiding complexity of having to understand what is `wp_template` vs `wp_template_part`. These could use single helper behind the scene ie. the one we have now. 

Or should we make it per view type and go with goToTemplateEditor({ id, type }) goToTemplatesList({ waitForActions, type }) with similar defaults we had now (`wp_template`)

Which one do you lean toward to more?

Something I know I'd like to keep is the ability to pass params/options as an object so that we can provide strong typing - passing by query string like we had before is too prone for errors.

sunyatasattva commented:
Totally fine with passing params via object. I'd differentiate per view type (second option) and perhaps instead of a `waitForActions` I'd have something like `waitFor: 'list' | 'anyEditedTemplate' = 'list'` or something along those lines, to make it more explicit what we are doing.

tomasztunik commented:
Sounds good - will update it quickly on Monday :)

tomasztunik committed:
improve clarity of the goToSiteEditor E2E util

- adds goToSiteEditor private helper that hides the temporary need
for handling theme.php and site-editor.php paths
- adds two exported utils goToTemplatesList and goToTemplateEdit
that hide UX details of waiting ofr page to be ready on load including special
case for waitingFor templates list actions to be loaded

tomasztunik committed:
put back page where it belongs ⚡️

sunyatasattva commented:
I'm loving it! Thanks for taking my feedback into account, Tomek! I'm now much more happy with this. Approved with flying colors! 🌈
